### PR TITLE
Remove more boost threads

### PR DESCRIPTION
**Pull request by theuni**
This doesn't completely get rid of boost::thread, but this batch should be easy to review, and leaves us with only threadGroup (scheduler + scriptcheck) remaining.

Note to reviewers: The upnp diff changes a bunch of whitespace, it's much more clear with 'git diff -w'

### CODE REVIEW

**Review by dcousens — APPROVED** [review]
light utACK

**Comment by practicalswift** [comment]
Very nice! Concept ACK!

File: src/qt/optionsmodel.cpp [hunk 1]
```diff
@@ -315,7 +315,11 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
             break;
         case MapPortUPnP: // core option - can be changed on-the-fly
             settings.setValue("fUseUPnP", value.toBool());
-            MapPort(value.toBool());
+            if (value.toBool()) {
+                StartMapPort();
+            } else {
+                StopMapPort();
```
Comment (laanwj) [inline src/qt/optionsmodel.cpp]: Probably call InterruptMapPort here too, before Stop, for consistency.

Edit: Oh, you have Stop call Interrupt? That's different from other [Start|Stop|Interrupt] constructions. As you probably know, the idea behind having a seperate interrupt is to have a two-stage (parallel) shutdown process, in which the interrupt can ask the subsystem to shut down, then Stop waits for it to really have stopped. Calling Stop without Interrupt could generally cause it to wait forever.

**Review by laanwj — COMMENTED** [review]
utACK

File: src/qt/optionsmodel.cpp (hunk 1 above)
Comment (theuni) [reply src/qt/optionsmodel.cpp]: Ok, I'll move it out.

Comment (laanwj) [reply src/qt/optionsmodel.cpp]: Sure - though if interrupt is idempotent there's no problem in calling it in Stop too.

**Comment by theuni** [comment]
Updated for @laanwj's suggestion, though I didn't see the second comment until I'd already pushed.

Diff from before is trivial:
```diff
diff --git a/src/net.cpp b/src/net.cpp
index cd9a5e1..2019146 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -1562,3 +1562,2 @@ void StopMapPort()
     if(g_upnp_thread.joinable()) {
-        InterruptMapPort();
         g_upnp_thread.join();
diff --git a/src/qt/optionsmodel.cpp b/src/qt/optionsmodel.cpp
index ee8ba65..909be1c 100644
--- a/src/qt/optionsmodel.cpp
+++ b/src/qt/optionsmodel.cpp
@@ -320,2 +320,3 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
             } else {
+                InterruptMapPort();
                 StopMapPort();
```

**Comment by laanwj** [comment]
reut-ACK https://github.com/bitcoin/bitcoin/pull/12381/commits/004f9999464c7ef4a57b281dcbb407e5d193e798, thanks

File: src/net.cpp [hunk 2]
```diff
@@ -1546,27 +1542,39 @@ void ThreadMapPort()
     }
 }
 
-void MapPort(bool fUseUPnP)
+void StartMapPort()
 {
-    static std::unique_ptr<boost::thread> upnp_thread;
+    if (!g_upnp_thread.joinable()) {
+        assert(!g_upnp_interrupt);
+        g_upnp_thread = std::thread((std::bind(&TraceThread<void (*)()>, "upnp", &ThreadMapPort)));
```
Comment (Empact) [inline src/net.cpp]: Nit: extra parens. :P

File: src/net.cpp [hunk 3]
```diff
@@ -1546,27 +1542,39 @@ void ThreadMapPort()
...
+
+void StopMapPort()
+{
+    if(g_upnp_thread.joinable()) {
+        g_upnp_thread.join();
+        g_upnp_interrupt.reset();
     }
 }
 
 #else
```
Comment (Empact) [inline src/net.cpp]: Maybe worth commenting that this is the USE_UPNP disabled case?

**Comment by theuni** [comment]
Since there's already an utACK, I'd prefer to leave @Empact's nits alone, but I'll be sure to include them if more changes are needed.

**Comment by jonasschnelli** [comment]
utACK 004f9999464c7ef4a57b281dcbb407e5d193e798

**Pull request merged**